analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs

```diff
@@ -26,24 +27,59 @@
 using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Common;
 using SonarAnalyzer.Helpers;
+using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
     public sealed class CastShouldNotBeDuplicated : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S3247";
-        private const string MessageFormat = "Replace this type-check-and-cast sequence with an 'as' and a null check.";
+        private const string DiagnosticId = "S3247";
+        private const string MessageFormat = "{0}";
+        private const string ReplaceWithAsAndNullCheckMessage = "Replace this type-check-and-cast sequence with an 'as' and a null check.";
+        private const string RemoveRedundantCaseMessage = "Remove this redundant cast.";
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
-        protected override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterSyntaxNodeActionInNonGenerated(CheckForIssue, SyntaxKind.IsExpression);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-        private void CheckForIssue(SyntaxNodeAnalysisContext analysisContext)
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(IsExpression, SyntaxKind.IsExpression);
+            context.RegisterSyntaxNodeActionInNonGenerated(IsPatternExpression, SyntaxKindEx.IsPatternExpression);
+        }
+
+        private static void IsPatternExpression(SyntaxNodeAnalysisContext analysisContext)
+        {
+            var isPatternExpression = (IsPatternExpressionSyntaxWrapper)analysisContext.Node;
+            if (!(isPatternExpression.SyntaxNode.GetFirstNonParenthesizedParent() is IfStatementSyntax parentIfStatement))
+            {
+                return;
+            }
+            var isPatternLocation = isPatternExpression.SyntaxNode.GetLocation();
+            if (isPatternExpression.Pattern.SyntaxNode.IsKind(SyntaxKindEx.RecursivePattern))
+            {
+                var recursivePattern = (RecursivePatternSyntaxWrapper)isPatternExpression.Pattern.SyntaxNode;
```

Comment:
As far as I can tell we handle only the first level of the recursive pattern. 
I think we should either support imbrication or document the FN.
e.g.
```
void M(object x, object y)
    {
        if ((x, y) is ((int a, int b), string v3))
        {
            
        }
    }
```

Reply:
Will document it as FN.